src/Mocking.jl

```diff
@@ -34,125 +37,57 @@ function enable(; force::Bool=false)
 end
 
 struct Patch
-    signature::Expr
-    body::Function
-    modules::Set
-    # translation::Dict
-
-    function Patch(signature::Expr, body::Function, translation::Dict)
-        trans = adjust_bindings(translation)
-        sig = name_parameters(absolute_signature(signature, trans))
-
-        # On VERSION >= v"0.5"
-        # modules = Set(b.args[1] for b in values(trans) if isa(b, Expr))
-        modules = Set()
-        for b in values(trans)
-            if isa(b, Expr)
-                push!(modules, b.args[1])
-            end
-        end
-
-        new(sig, body, modules)
-    end
-end
-
-# TODO: Find non-eval way to determine module locations of Types
-# evaling in the @patch scope seems to be problematic for pre-compliation
-# first(methods(x)).sig.types[2:end]
-
-# We can use the @patch macro to create a list of bindings used then pass that
-# in as an array into Patch. At runtime the types and function names will be fully
-# qualified
-
-# We can support optional parameters and keywords by using generic functions on
-# 0.4
-
-function convert(::Type{Expr}, p::Patch)
-    exprs = Expr[]
-
-    # Generate imports for all required modules
-    for m in p.modules
-        bindings = splitbinding(m)
-
-        :Main in bindings && error("Mocking cannot handle bindings from Main.")
-
-        for i in 1:length(bindings)
-            import_expr = Expr(:import, Expr(:., bindings[1:i]...))
-            push!(exprs, import_expr)
-        end
-    end
-
-    # Generate the new method which will call the user's patch function. We need to perform
-    # this call instead of injecting the body expression to support closures.
-    sig, body = p.signature, p.body
-    params = call_parameters(sig)
-    push!(exprs, Expr(:(=), sig, Expr(:block, Expr(:call, body, params...))))
-
-    return Expr(:block, exprs...)
+    target::Callable
+    alternate::Function
 end
 
 macro patch(expr::Expr)
-    if expr.head == :function
-        name = expr.args[1].args[1]
-        params = expr.args[1].args[2:end]
-        body = expr.args[2]
-
-    # Short-form function syntax
-    elseif expr.head == :(=) && expr.args[1].head == :call
-        name = expr.args[1].args[1]
+    # Expect a named function in long-form or short-form
+    if expr.head === :function || (expr.head === :(=) && expr.args[1].head == :call)
+        target_name = expr.args[1].args[1]
         params = expr.args[1].args[2:end]
         body = expr.args[2]
-
-    # Anonymous function syntax
-    # elseif expr.head == :(->)
-        # TODO: Determine how this could be supported
+    elseif expr.head == :(->)
+        throw(ArgumentError("expression needs to be a named function"))
     else
         throw(ArgumentError("expression is not a function definition"))
     end
```

Comment:
How do you feel about ising `SyntaxError` for this?

Reply:
Isn't that an internal Julia error? I'm not sure I can construct an instance of that type

Reply:
Apparently so.  I feel like I used to do this.